tpm2tools/seal_test.go

```diff
@@ -0,0 +1,238 @@
+	pcrList := []int{1, 2, 3, 4, 5, 6, 7}
+	extension := bytes.Repeat([]byte{0xAA}, sha256.Size)
+
+	for _, pcrNum := range pcrList {
+		err = tpm2.PCRExtend(rwc, tpmutil.Handle(pcrNum), tpm2.AlgSHA256, extension, "")
+		if err != nil {
+			t.Fatalf("failed to extend pcr: %v", err)
+		}
+	}
+
+	sealed, err := key.Seal(rwc, pcrList, secret)
+	if err != nil {
+		t.Fatalf("failed to seal: %v", err)
+	}
+
+	unseal, err := key.Unseal(rwc, pcrList, sealed)
+	if err != nil {
+		t.Fatalf("failed to unseal: %v", err)
+	}
+	if !bytes.Equal(secret, unseal) {
+		t.Fatalf("unsealed (%v) not equal to secret (%v)", unseal, secret)
+	}
+
+	err = tpm2.PCRExtend(rwc, tpmutil.Handle(7), tpm2.AlgSHA256, extension, "")
+	if err != nil {
+		t.Fatalf("failed to extend pcr: %v", err)
+	}
+
+	// unseal should not succeed.
+	unseal, err = key.Unseal(rwc, pcrList, sealed)
+	if err == nil {
+		t.Fatalf("unseal should have caused an error: %v", err)
+	}
+}
+
+func TestComputeSessionAuth(t *testing.T) {
+	rwc := internal.GetTPM(t)
+	defer rwc.Close()
+
+	pcrList := []int{1, 2, 3, 4, 5, 6, 7}
+
+	pcrs := map[int][]byte{}
+
+	for _, pcrNum := range pcrList {
+		pcrVal, err := tpm2.ReadPCR(rwc, pcrNum, tpm2.AlgSHA256)
+		if err != nil {
+			t.Fatalf("failed to read pcr: %v", err)
+		}
+
+		pcrs[pcrNum] = pcrVal
+	}
+
+	pcrNums := make([]int, 0, len(pcrs))
+	for key := range pcrs {
+		pcrNums = append(pcrNums, key)
+	}
+
+	getAuth, err := getPCRSessionAuth(rwc, pcrNums)
+	if err != nil {
+		t.Fatalf("failed to get session auth: %v", err)
+	}
+
+	computeAuth, err := computePCRSessionAuth(pcrs)
+	if err != nil {
+		t.Fatalf("failed to compute session auth: %v", err)
+	}
+
+	if !bytes.Equal(computeAuth, getAuth) {
+		t.Fatalf("computed auth (%v) not equal to session auth(%v)", computeAuth, getAuth)
+	}
+}
+
+func TestSelfReseal(t *testing.T) {
+	rwc := internal.GetTPM(t)
+	defer rwc.Close()
+
+	key, err := StorageRootKeyRSA(rwc)
+	if err != nil {
+		t.Fatalf("can't create srk from template: %v", err)
+	}
+
+	secret := []byte("test")
+
+	pcrList := []int{1, 2, 3, 4, 5, 6, 7}
+	pcrs := map[int][]byte{}
+	for _, pcrNum := range pcrList {
+		pcrVal, err := tpm2.ReadPCR(rwc, pcrNum, tpm2.AlgSHA256)
+		if err != nil {
+			t.Fatalf("failed to read pcr: %v", err)
+		}
+
+		pcrs[pcrNum] = pcrVal
+	}
+
+	sealed, err := key.Seal(rwc, pcrList, secret)
+	if err != nil {
+		t.Fatalf("failed to seal: %v", err)
+	}
+
+	unseal, err := key.Unseal(rwc, pcrList, sealed)
+	if err != nil {
+		t.Fatalf("failed to unseal: %v", err)
+	}
+	if !bytes.Equal(secret, unseal) {
+		t.Fatalf("unsealed (%v) not equal to secret (%v)", unseal, secret)
+	}
+
+	sealed, err = key.Reseal(rwc, pcrs, sealed)
+	if err != nil {
+		t.Fatalf("failed to reseal: %v", err)
+	}
+
+	unseal, err = key.Unseal(rwc, pcrList, sealed)
+	if err != nil {
+		t.Fatalf("unseal failed: %v", err)
+	}
+}
+
+func computePCRValue(t *testing.T, extensions [][]byte) ([]byte, error) {
+	rwc := internal.GetTPM(t)
+	defer rwc.Close()
+
+	pcrNum := 0
```

Comment:
I'm confused here, why does this use TPM commands at all? Can't we just do this extension just using go's sha256 hash?

Reply:
I was trying to avoid implementing hash extension in this patch, as I mentioned in the previous thread, to keep this from patch from ballooning too much.

This is the third ambiguous comment here. If you want a change please state as much.